### PR TITLE
Improve Job integration tests runtime by reusing apiserver within each test

### LINKED ISSUE #126225: Investigate improving Job integration test runtime by reusing the api server
This is a spin off from https://github.com/kubernetes/kubernetes/issues/126202 as there are still some tests which spawn api server per test case (which takes around 2s). Example https://github.com/kubernetes/kubernetes/blob/dda657b5982e8f9102b8df5931344262b0793163/test/integration/job/job_test.go#L431-L432

The idea is to spawn a single API server, and to avoid Job / pod name collisions use a unique suffix per test case (say test case index).

/sig apps

### PR DESCRIPTION
**Pull request by kaisoz**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

As explained in [this issue](https://github.com/kubernetes/kubernetes/issues/126202), integration tests are very slow at the moment. This PR improves job integration tests execution time by spawning a single api server per test case as suggested in #126225

With these changes, the execution time of the test suite went down from [14m16s](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126191/pull-kubernetes-integration/1813929954224640000) to [6m57s](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126439/pull-kubernetes-integration/1833134937067229184), which is a **decrease of ~ 51.29%**

#### Which issue(s) this PR fixes:

Fixes #126225

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by mimowo** [comment]
@kaisoz please update the PR description summarizing the performance gains per test

File: test/integration/job/job_test.go [hunk 1]
```diff
@@ -4475,11 +4523,7 @@ func removePodsFinalizer(ctx context.Context, t *testing.T, clientSet clientset.
 		t.Fatalf("Failed to list pods: %v", err)
 	}
 	updatePod(ctx, t, clientSet, pods.Items, func(pod *v1.Pod) {
-		for i, finalizer := range pod.Finalizers {
```
Comment (dejanzele) [inline test/integration/job/job_test.go:4478]: What does this change do?

Comment (kaisoz) [reply test/integration/job/job_test.go:4478]: Sorry @dejanzele , I didn't ping you yet because this is still WIP. I got problems with a couple of tests which I discussed with @mimowo on Slack. I'm currently working in the fix (which will outdate this change). I'll assign you as soon as I push the changes.

**Comment by kaisoz** [comment]
/assign @dejanzele 

I would appreciate if you could have a look while the tests run (they do pass locally). Once they are green I'll update the description with the performance gains 😊 

Thanks!

File: test/integration/job/job_test.go [hunk 2]
```diff
@@ -3477,8 +3525,8 @@ func BenchmarkLargeFailureHandling(b *testing.B) {
 	}
 }
 
-// cleanUp deletes all pods and the job
-func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1.Job) error {
+// cleanUp deletes all pods and the job. Waits for the job deletion if sync is true
```
Comment (dejanzele) [inline test/integration/job/job_test.go:3528]:
```suggestion
// cleanUp deletes all pods and the job and waits for the job deletion if sync is set to true.
```

File: test/integration/job/job_test.go [hunk 3]
```diff
@@ -1520,16 +1531,18 @@ func TestDelayTerminalPhaseCondition(t *testing.T) {
...
 			t.Cleanup(cancel)
 
+			test.job.Name = fmt.Sprintf("test-job-%d", i)
 			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &test.job)
 			if err != nil {
 				t.Fatalf("Error %q while creating the job %q", err, jobObj.Name)
 			}
-			t.Cleanup(func() { removePodsFinalizer(ctx, t, clientSet, ns.Name) })
+			t.Cleanup(func() {
+				removePodsFinalizerByName(ctx, t, clientSet, ns.Name, "fake.example.com/blockDeletion")
```
Comment (dejanzele) [inline test/integration/job/job_test.go:1543]: @mimowo is it worth the extra effort to remove multiple finalizers in a single call?
Something like:
```
removePodsFinalizersByName(ctx, t, clientSet, ns.Name, []string{"fake.example.com/blockDeletion", "batch.kubernetes.io/job-tracking"})
```

File: test/integration/job/job_test.go [hunk 4]
```diff
@@ -2061,20 +2078,33 @@ func TestManagedBy(t *testing.T) {
...
 			ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
 			defer cancel()
+
+			test.job.Name = fmt.Sprintf("test-job-%d", i)
 			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &test.job)
 			if err != nil {
 				t.Fatalf("Error %v while creating the job %q", err, klog.KObj(jobObj))
 			}
 
+			defer func() {
```
Comment (dejanzele) [inline test/integration/job/job_test.go:2099]: `t.Cleanup` might be a better approach instead of `defer`.

Here's an interesting article on the differences between `t.Cleanup` and `defer` - https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

File: test/integration/job/job_test.go [hunk 5]
```diff
@@ -3498,14 +3546,40 @@ func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1
...
+		return err
+	}
+	if sync {
+		err := wait.PollUntilContextTimeout(ctx, waitInterval, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+			_, err := clientSet.BatchV1().Jobs(jobObj.Namespace).Get(ctx, jobObj.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return true, nil
+				} else {
+					return true, fmt.Errorf("Error checking the deleted job: %w", err)
```
Comment (dejanzele) [inline test/integration/job/job_test.go:3565]:
```suggestion
					return true, fmt.Errorf("Error getting the deleted job: %w", err)
```

File: test/integration/job/job_test.go [hunk 6]
```diff
@@ -3498,14 +3546,40 @@ func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1
...
+	if err != nil {
+		return err
+	}
+	if sync {
+		err := wait.PollUntilContextTimeout(ctx, waitInterval, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+			_, err := clientSet.BatchV1().Jobs(jobObj.Namespace).Get(ctx, jobObj.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return true, nil
+				} else {
```
Comment (dejanzele) [inline test/integration/job/job_test.go:3564]: Else block is reduntant here, it can safely be outdented as the if clause has a return

File: test/integration/job/job_test.go [hunk 7]
```diff
@@ -3498,14 +3546,40 @@ func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1
...
+
+	// Set the propagation policy to background to ensure that the job doesn't receive any finalizer at deletion time
+	backgroundDeletion := metav1.DeletePropagationBackground
+	err := clientSet.BatchV1().Jobs(jobObj.Namespace).Delete(ctx, jobObj.Name, metav1.DeleteOptions{
+		PropagationPolicy: &backgroundDeletion,
+	})
+	if err != nil {
+		return err
+	}
+	if sync {
```
Comment (dejanzele) [inline test/integration/job/job_test.go:3558]: It might be cleaner if it checked the reverse condition:
```
if !sync {
  return nil
}
```

and the whole sync logic can be outdented

File: test/integration/job/job_test.go (hunk 5 above)
Comment (dejanzele) [reply test/integration/job/job_test.go:3565]: Actually, we might even not need to wrap the error here.

File: test/integration/job/job_test.go (hunk 2 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3528]: mm and what do you think of

```suggestion
// cleanUp deletes all pods and the job, then waits for job deletion if sync is set to true.
```

File: test/integration/job/job_test.go (hunk 4 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:2099]: Makes sense to me, and if there's no objection I'd replace every `defer` to `t.Cleanup` 👍🏻

File: test/integration/job/job_test.go (hunk 3 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:1543]: I don't have a strong opinion about this, it came out of a discussion with @mimowo on Slack. You could argue that having different sequential calls makes more explicit to the reader the order in which the finalizers should be removed, but it's a brittle argument...

**Comment by kaisoz** [comment]
Hey @dejanzele ! Thanks for the review 😊 I turns out that the `sync cleanUp` might not be needed because it seems that the same can be achieved by using only `DeletePropagationBackground` policy. I just need to **confirm** that this **guarantees** that the job is deleted before the next test executes.

File: test/integration/job/job_test.go (hunk 7 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3558]: The sync logic might not be needed at all, see this [comment](https://github.com/kubernetes/kubernetes/pull/126439#issuecomment-2272139118)

File: test/integration/job/job_test.go (hunk 6 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3564]: The sync logic might not be needed at all, see this [comment](https://github.com/kubernetes/kubernetes/pull/126439#issuecomment-2272139118).

File: test/integration/job/job_test.go (hunk 5 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3565]: The sync logic might not be needed at all, see this [comment](https://github.com/kubernetes/kubernetes/pull/126439#issuecomment-2272139118)

File: test/integration/job/job_test.go (hunk 3 above)
Comment (mimowo) [reply test/integration/job/job_test.go:1543]: > @mimowo is it worth the extra effort to remove multiple finalizers in a single call? Something like:
> 
> ```
> removePodsFinalizersByName(ctx, t, clientSet, ns.Name, []string{"fake.example.com/blockDeletion", "batch.kubernetes.io/job-tracking"})
> ```

I like the proposal.

Comment (mimowo) [reply test/integration/job/job_test.go:1543]: > You could argue that having different sequential calls makes more explicit to the reader the order in which the finalizers should be removed 

I was only concerned with doing `Finalizers = nil`, as this could hide some potentially important details.

File: test/integration/job/job_test.go (hunk 4 above)
Comment (mimowo) [reply test/integration/job/job_test.go:2099]: Thanks for pointing it out. I'm happy to migrate to `t.Cleanup`, but if there are too many unrelated changes, then I would suggest a separate PR.

File: test/integration/job/job_test.go [hunk 8]
```diff
@@ -424,17 +424,20 @@ func TestJobPodFailurePolicy(t *testing.T) {
 			},
 		},
 	}
+
+	i := 0
+	closeFn, restConfig, clientSet, ns := setup(t, "simple")
```
Comment (mimowo) [inline test/integration/job/job_test.go:429]:
```suggestion
	closeFn, restConfig, clientSet, ns := setup(t, "pod-failure-policy")
```
nit, but as we already touch this line it makes sense to adjust. I guess it might be helpful for debugging when reading logs.

File: test/integration/job/job_test.go [hunk 9]
```diff
@@ -758,18 +762,20 @@ func TestSuccessPolicy(t *testing.T) {
 			wantConditionTypes: []batchv1.JobConditionType{batchv1.JobFailed},
 		},
 	}
+
+	i := 0
+	closeFn, restConfig, clientSet, ns := setup(t, "simple")
```
Comment (mimowo) [inline test/integration/job/job_test.go:767]:
```suggestion
	closeFn, restConfig, clientSet, ns := setup(t, "success-policy")
```
nit

File: test/integration/job/job_test.go [hunk 10]
```diff
@@ -1898,17 +1912,19 @@ func TestBackoffLimitPerIndex(t *testing.T) {
 			},
 		},
 	}
+
+	i := 0
+	closeFn, restConfig, clientSet, ns := setup(t, "simple")
```
Comment (mimowo) [inline test/integration/job/job_test.go:1917]:
```suggestion
	closeFn, restConfig, clientSet, ns := setup(t, "backoff-limit-per-index")
```

File: test/integration/job/job_test.go (hunk 7 above)
Comment (mimowo) [reply test/integration/job/job_test.go:3558]: It would be cool to drop the sync, the [docs](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#background-deletion) say: "In background cascading deletion, the Kubernetes API server deletes the owner object **immediately** and the controller cleans up the dependent objects in the background. ".
So, IIUC this is sync for the Job object already.

**Review by mimowo — COMMENTED** [review]
Generally LGTM. Indeed, it would be nice to eliminate the `sync` from `cleanUp` if not needed.

File: test/integration/job/job_test.go (hunk 3 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:1543]: ok! so I'll implement it 😊

**Comment by kaisoz** [comment]
So as @mimowo [says](https://github.com/kubernetes/kubernetes/pull/126439#discussion_r1706467325), the docs state that with background propagation policy the object gets deleted immediately and the dependent objects are deleted in the background. Besides, according to the code, the default finalizers are only added when the propagation policy is not background.

So I've finally removed the sync `cleanUp` and tests keep passing.

File: test/integration/job/job_test.go (hunk 7 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3558]: I finally dropped the sync code, see [this comment](https://github.com/kubernetes/kubernetes/pull/126439#issuecomment-2275307459) 😊

File: test/integration/job/job_test.go (hunk 6 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3564]: I finally dropped the sync code, see [this comment](https://github.com/kubernetes/kubernetes/pull/126439#issuecomment-2275307459) 😊

File: test/integration/job/job_test.go (hunk 5 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3565]: I finally dropped the sync code, see [this comment](https://github.com/kubernetes/kubernetes/pull/126439#issuecomment-2275307459) 😊

File: test/integration/job/job_test.go [hunk 11]
```diff
@@ -1520,16 +1529,17 @@ func TestDelayTerminalPhaseCondition(t *testing.T) {
...
 			t.Cleanup(cancel)
 
+			test.job.Name = fmt.Sprintf("test-job-%d", i)
 			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &test.job)
 			if err != nil {
 				t.Fatalf("Error %q while creating the job %q", err, jobObj.Name)
 			}
-			t.Cleanup(func() { removePodsFinalizer(ctx, t, clientSet, ns.Name) })
+			t.Cleanup(func() {
+				removePodsFinalizersByName(ctx, t, clientSet, ns.Name, []string{"fake.example.com/blockDeletion", "batch.kubernetes.io/job-tracking"})
```
Comment (mimowo) [inline test/integration/job/job_test.go:1541]: nit: I would suggest to declare the custom finalizer in the test, say `blockDeletionFinalizerForTest`. For the Job finalizer we can use the API const `JobTrackingFinalizer`.

**Review by mimowo — COMMENTED** [review]
LGTM, just left a minor comment. Getting the time below 10min again is great!

File: test/integration/job/job_test.go (hunk 4 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:2100]: Done, I've migrated to `t.Cleanup` only the tests I had to modified and leaved the rest for a future cleanup PR 👍🏻

File: test/integration/job/job_test.go [hunk 12]
```diff
@@ -2061,20 +2079,32 @@ func TestManagedBy(t *testing.T) {
...
 			if err != nil {
 				t.Fatalf("Error %v while creating the job %q", err, klog.KObj(jobObj))
 			}
 
+			defer func() {
+				removePodsFinalizersByName(ctx, t, clientSet, ns.Name, []string{blockDeletionFinalizerForTest, batchv1.JobTrackingFinalizer})
+				// Wait for cleanup to finish to prevent this job from affecting metrics
+				if err := cleanUp(ctx, clientSet, jobObj); err != nil {
+					t.Fatalf("Failed cleanup: %v", err)
+				}
```
Comment (atiratree) [inline test/integration/job/job_test.go:2105]: Why is this needed now? Only for the correct metric counting?

And why is it not present in other places like `TestJobPodFailurePolicy`?

File: test/integration/job/job_test.go [hunk 13]
```diff
@@ -2061,20 +2079,32 @@ func TestManagedBy(t *testing.T) {
...
 			defer cancel()
+
+			test.job.Name = fmt.Sprintf("test-job-%d", i)
 			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &test.job)
 			if err != nil {
 				t.Fatalf("Error %v while creating the job %q", err, klog.KObj(jobObj))
 			}
 
+			defer func() {
+				removePodsFinalizersByName(ctx, t, clientSet, ns.Name, []string{blockDeletionFinalizerForTest, batchv1.JobTrackingFinalizer})
```
Comment (atiratree) [inline test/integration/job/job_test.go:2101]: might be worth passing the finalizers straight to the cleanUp function

File: test/integration/job/job_test.go (hunk 12 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:2105]: Exactly! This particular test was failing because the `jobs_by_external_controller_total` metric included jobs from previous test cases due to the reuse of the API server. That's why I extended the `cleanUp` function to ensure that the job is also deleted and call it on each test end.

I don't call the function in other places because the presence of old the job/pods don't affect the test results and I considered that there's no need to add extra API calls since they'd add up to the runtime.

File: test/integration/job/job_test.go (hunk 13 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:2101]: I kept the finalizers removal in a separated function because some tests where already doing this (see [this example](https://github.com/kubernetes/kubernetes/pull/126439/files#diff-ace35385ec6f5400e8ccbfe8e9f35c2f9573d0fc05889071dc9c4b36e5e975f7R1544)). And since they were running fine, I didn't want to add extra unnecessary calls.

However, if you prefer so, I can implement your suggestion and see how it all behaves.

Comment (atiratree) [reply test/integration/job/job_test.go:2101]: I think it would be more readable that way. We can pass the finalizers to the cleanup function. 

> (see [this example](https://github.com/kubernetes/kubernetes/pull/126439/files#diff-ace35385ec6f5400e8ccbfe8e9f35c2f9573d0fc05889071dc9c4b36e5e975f7R1544)).

Can we also call cleanup in this place? this way we do not have to check how many and which functions we have to call to do a proper cleanup in each place.

File: test/integration/job/job_test.go [hunk 14]
```diff
@@ -3443,7 +3491,8 @@ func BenchmarkLargeFailureHandling(b *testing.B) {
 					b.Fatalf("Failed to create Job: %v", err)
 				}
 				b.Cleanup(func() {
-					if err := cleanUp(ctx, clientSet, jobObj); err != nil {
+					// TODO
```
Comment (mimowo) [inline test/integration/job/job_test.go:3494]: nit: drop this TODO.

File: test/integration/job/job_test.go [hunk 15]
```diff
@@ -3498,14 +3550,23 @@ func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1
...
-	return clientSet.BatchV1().Jobs(jobObj.Namespace).Delete(ctx, jobObj.Name, metav1.DeleteOptions{})
+
+	// Set the propagation policy to background to ensure that the job doesn't receive any finalizer at deletion time
+	backgroundDeletion := metav1.DeletePropagationBackground
+	if err := clientSet.BatchV1().Jobs(jobObj.Namespace).Delete(ctx, jobObj.Name, metav1.DeleteOptions{
+		PropagationPolicy: &backgroundDeletion,
+	}); err != nil {
+		return err
+	}
+	return nil
```
Comment (mimowo) [inline test/integration/job/job_test.go:3561]:
```suggestion
	return clientSet.BatchV1().Jobs(jobObj.Namespace).Delete(ctx, jobObj.Name, metav1.DeleteOptions{
		PropagationPolicy: ptr.To(metav1.DeletePropagationBackground),
	})
```
nit

File: test/integration/job/job_test.go [hunk 16]
```diff
@@ -3498,14 +3550,23 @@ func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1
...
 }
 
 func TestOrphanPodsFinalizersClearedWithGC(t *testing.T) {
+	i := 0
+	closeFn, restConfig, clientSet, ns := setup(t, "simple")
```
Comment (mimowo) [inline test/integration/job/job_test.go:3566]:
```suggestion
	closeFn, restConfig, clientSet, ns := setup(t, "orphan-pod-finalizers")
```

File: test/integration/job/job_test.go (hunk 14 above)
Comment (kaisoz) [reply test/integration/job/job_test.go:3494]: 🤦🏻 done

File: test/integration/job/job_test.go (hunk 13 above)
Comment (kaisoz) [reply test/integration/job/job_test.go]: @atiratree I implemented your suggestions, please have a look whenever you have time 😊

File: test/integration/job/job_test.go [hunk 17]
```diff
@@ -2061,20 +2081,31 @@ func TestManagedBy(t *testing.T) {
...
+	t.Cleanup(closeFn)
 	for name, test := range testCases {
 		t.Run(name, func(t *testing.T) {
 			resetMetrics()
 			featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobManagedBy, test.enableJobManagedBy)
 
-			closeFn, restConfig, clientSet, ns := setup(t, "managed-by")
-			defer closeFn()
 			ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
 			defer cancel()
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:2094]: use `t.Cleanup`

File: test/integration/job/job_test.go [hunk 18]
```diff
@@ -4468,23 +4533,28 @@ func deletePods(ctx context.Context, t *testing.T, clientSet clientset.Interface
...
 	pods, err := clientSet.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
 	if err != nil {
-		t.Fatalf("Failed to list pods: %v", err)
+		return err
+	}
+
+	finalizersSet := make(map[string]struct{}, len(finalizersNames))
+	for _, finalizer := range finalizersNames {
+		finalizersSet[finalizer] = struct{}{}
 	}
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:4545]:
```suggestion
	finalizersSet := sets.New(finalizersNames...)
```

File: test/integration/job/job_test.go [hunk 19]
```diff
@@ -424,17 +424,20 @@ func TestJobPodFailurePolicy(t *testing.T) {
...
-			closeFn, restConfig, clientSet, ns := setup(t, "simple")
-			defer closeFn()
 			ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
-			defer func() {
+			t.Cleanup(func() {
 				cancel()
-			}()
+			})
 
+			test.job.Name = fmt.Sprintf("test-job-%d", i)
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:440]: What is the goal of having a different name?

Because subtests will run in different order every time, this `i` is not predictable or it doesn't help much when debugging. Maybe it's simpler to use `test.job.GenerateName` to generate a random string.

File: test/integration/job/job_test.go (hunk 13 above)
Comment (atiratree) [reply test/integration/job/job_test.go]: nice!

File: test/integration/job/job_test.go [hunk 20]
```diff
@@ -4468,23 +4498,25 @@ func deletePods(ctx context.Context, t *testing.T, clientSet clientset.Interface
 	}
 }
 
-func removePodsFinalizer(ctx context.Context, t *testing.T, clientSet clientset.Interface, namespace string) {
-	t.Helper()
+func removePodsFinalizersByName(ctx context.Context, clientSet clientset.Interface, namespace string, finalizersNames []string) error {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply test/integration/job/job_test.go:4501]: +1

[the remark this answers is not in the document]
Comment (kaisoz) [reply test/integration/job/job_test.go]: sure! makes sense. Done!

**Comment by mimowo** [comment]
/lgtm
/retitle Improve Job integration tests runtime by reusing apiserver within each test
Suggestion, so that one gets some understanding of what is changed just by the title.
